### PR TITLE
HPCC-20955 Convert Topology and related XSDs to new format

### PR DESCRIPTION
**Pull request by kenrowland**
Convert XSDs to new format

Signed-off-by: Ken Rowland <kenneth.rowland@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Review by RussWhitehead — COMMENTED** [review]
Only a few comments on the CPP files, I defer the XSD to someone else

File: configuration/configmgr/configmgrlib/EnvironmentValue.cpp [hunk 1]
```diff
@@ -121,16 +121,17 @@ void EnvironmentValue::initialize()
         //
         // type "prefix" means to use the auto generate value as a name prefix and to append numbers until a new unique name is
         // found. ("prefix_" is a variation that adds an underbar (_) when appending numbers)
```
[the remark this answers is not in the document]
Comment (kenrowland) [reply configuration/configmgr/configmgrlib/EnvironmentValue.cpp]: It has been documented in the Wiki. See http://10.240.34.10/wiki/index.php/XSD_Schema_Definition#Automatic_Value_Generation

File: configuration/configmgr/configmgrlib/SchemaValue.hpp [hunk 2]
```diff
@@ -123,6 +125,7 @@ class DECL_EXPORT SchemaValue
             unsigned m_deprecated: 1;
             unsigned m_isUnique  : 1;
             unsigned m_isDefined : 1;
+            unsigned m_noOutput  : 1;
```
[the remark this answers is not in the document]
Comment (kenrowland) [reply configuration/configmgr/configmgrlib/SchemaValue.hpp]: It was a request from Tony to use a bit field to reduce memory usage. Performance hits are negligible. Alignment is not an issue. These values are in lieu of full bool flags.

File: configuration/configmgr/configmgrlib/XMLEnvironmentMgr.cpp [hunk 3]
```diff
@@ -75,7 +75,7 @@ void XMLEnvironmentMgr::serialize(pt::ptree &envTree, std::shared_ptr<Environmen
         std::string attrValue;
         attrValue = (*attrIt)->getValue();
 
-        if (!attrValue.empty())
+        if (!attrValue.empty() && !(*attrIt)->getSchemaValue()->isNoOutput())
```
[the remark this answers is not in the document]
Comment (kenrowland) [reply configuration/configmgr/configmgrlib/XMLEnvironmentMgr.cpp]: Not likely. The reason for a noOutput attribute is to present information only to the user. It is not desired to have the value written to the environment.

File: initfiles/componentfiles/configschema/xsd/eclagent.xsd [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+-->
+
+<xs:schema
+        xmlns:xs="http://www.w3.org/2001/XMLSchema" elementFormDefault="qualified" attributeFormDefault="unqualified"
```
[the remark this answers is not in the document]
Comment (kenrowland) [reply initfiles/componentfiles/configschema/xsd/eclagent.xsd]: Can you provide further information as to your concern about indent here and other xsd?

**Review by kenrowland — COMMENTED** [review]
@RussWhitehead Please see my responses

File: configuration/configmgr/configmgrlib/EnvironmentValue.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (RussWhitehead) [reply configuration/configmgr/configmgrlib/EnvironmentValue.cpp]: A single line comment in the code is better than a developer having to read/search a Wiki. Please add

File: initfiles/componentfiles/configschema/xsd/eclagent.xsd (hunk 4 above)
[the remark this answers is not in the document]
Comment (RussWhitehead) [reply initfiles/componentfiles/configschema/xsd/eclagent.xsd]: Never mind

File: configuration/configmgr/configmgrlib/SchemaValue.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (RussWhitehead) [reply configuration/configmgr/configmgrlib/SchemaValue.hpp]: Certainly it doesn't matter here, but I was curious why a bitfield is better than a struct of bools. If memory usage is of concern, then yes this is an appropriate usage. Leave as is

**Comment by kenrowland** [comment]
@RussWhitehead Please see latest commit

**Review by RussWhitehead — APPROVED** [review]
@kenrowland Looks great

**Comment by kenrowland** [comment]
@richardkchapman Please merge

**Pull request merged**